version: "3"

services:
  namenode:
    build: ./Namenode
    hostname: namenode
    container_name: namenode
    volumes:
      - hadoop_namenode:/hadoop/dfs/name
      - ./data/:/hadoop-data/input
      - ./map_reduce/:/hadoop-data/map_reduce
    environment:
      - CLUSTER_NAME=test
    ports:
      - "9870:9870"
      - "8020:8020"
    networks:
      - hadoop_network

  resourcemanager:
    build: ./Resourcemanager
    container_name: resourcemanager
    hostname: resourcemanager
    restart: on-failure
    depends_on:
      - namenode
      - datanode1
      - datanode2
    ports:
      - "8089:8088"
    networks:
      - hadoop_network

  nodemanager1:
    build: ./Nodemanager
    hostname: nodemanager1
    container_name: nodemanager1
    depends_on:
      - namenode
      - datanode1
      - datanode2
    ports:
      - "8042:8042"
    networks:
      - hadoop_network

  nodemanager2:
    build: ./Nodemanager
    hostname: nodemanager2
    container_name: nodemanager2
    depends_on:
      - namenode
      - datanode1
      - datanode2
    ports:
      - "8043:8042"
    networks:
      - hadoop_network

  datanode1:
    build: ./Datanode
    hostname: datanode1
    container_name: datanode1
    depends_on:
      - namenode
    volumes:
      - hadoop_datanode1:/hadoop/dfs/data
    networks:
      - hadoop_network

  datanode2:
    build: ./Datanode
    hostname: datanode2
    container_name: datanode2
    depends_on:
      - namenode
    volumes:
      - hadoop_datanode2:/hadoop/dfs/data
    networks:
      - hadoop_network

volumes:
  hadoop_namenode:
  hadoop_datanode1:
  hadoop_datanode2:


networks:
  hadoop_network:
    name: hadoop_network
